feat: transitioning the hiding circuits to Noir completely

kashbrti opened the pull request:
This is the last PR in the thread of the work migrating the hiding kernel(circuit) to Noir. 
This PR: 
- adds the Noir implementation of the hiding kernels to noir-protocol-circuits 
- adds the mock hiding kernel for generating the vks (in noir-protocol-circuits)
- changes the PXE flow to run the new hiding circuit 
(thanks to @LeilaWang for all the above) 
- adds the handling of the hiding circuit coming from Noir in `bbapi`
- modifies the api-client-ivc flow slightly to accommodate the new changes 
- mimics the changes in bbapi in `backend.ts`

Commit:
[empty] Start merge-train. Choo choo.

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

LeilaWang committed:
Hiding circuits.

LeilaWang committed:
Update proof type constant.

Commit:
[empty] Start merge-train. Choo choo.

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
feat: remove duplication of hiding circuit creation by handling its completion outside of ClientIVC::prove (#16135)

This PR changes the way the construction of the hiding kernel is handled
to accommodate for the hiding kernel being implemented in Noir.
Changes include: 
- added two methods `construct_and_accumulate_next_circuit` and
`construct_hiding_circuit` to
`PrivateFunctionExecutionMockCircuitProducer`, to handle creation of the
hiding circuit in our testing modules
- removed the duplication of `complete_hiding_circuit_logic`. Now the
`hiding_circuit` builder is saved as a member `ClientIVC` class.
- added a `construct_and_accumulate_mock_kernel` method to
`ivc_recursion_constraints` test module.
- updated the API according to the changes to the `ClientIVC` class.

---------
Co-authored-by: Khashayar Barooti <khashayar@aztecprotocol.com>

Commit:
chore: remove is_kernel from MegaCircuitBuilder (#16186)

Remove `is_kernel` flag from the MegaCircuitBuilder as the functionality
it provides (knowing when we are building a kernel circuit) can be
handled solely in the ClientIVC class.

Commit:
Merge branch 'next' into merge-train/barretenberg

kashbrti committed:
Merge branch 'merge-train/barretenberg' into lw/hiding_circuit

kashbrti committed:
WIP

kashbrti committed:
wip

kashbrti committed:
wip

kashbrti committed:
changed the location of the expected vk being computed

kashbrti committed:
make the native tests pass

Commit:
[empty] Start merge-train. Choo choo.

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

ludamad committed:
chore(bb): properly check for churned vks (#16312)

Commit:
Merge branch 'next' into merge-train/barretenberg

Commit:
Merge branch 'next' into merge-train/barretenberg

kashbrti committed:
reverting changes

kashbrti committed:
removed unneeded files

kashbrti committed:
added the new prover tomls

kashbrti committed:
Merge branch 'merge-train/barretenberg' into kb/bbapi_changes_for_hiding_circuit

TomAFrench commented on noir-projects/noir-protocol-circuits/crates/hiding-kernel-to-rollup/src/main.nr:
```diff
@@ -0,0 +1,46 @@
+        // The associated public inputs and proofs are linked through the verification queue in the backend.
+        // The proof type `PROOF_TYPE_PG_FINAL` designates the final proof in the client IVC sequence.
+        if !std::runtime::is_unconstrained() {
+            std::verify_proof_with_type(
+                self.previous_kernel_vk_data.vk.key,
+                [],
+                [],
+                self.previous_kernel_vk_data.vk.hash,
+                PROOF_TYPE_PG_FINAL,
+            );
```
Looks like this should be in  https://github.com/AztecProtocol/aztec-packages/blob/next/barretenberg/noir/bb_proof_verification/src/lib.nr

kashbrti replied:
I will be moving all proof verification to use the library in a cleanup PR. I guess these methods will stay public anyways?

TomAFrench replied:
These methods being `verify_proof_with_type`? Yeah it'll remain public but we really don't want people to be accessing it directly if we can avoid it.

kashbrti committed:
changes to backend.ts

kashbrti committed:
added more things to bb ts

kashbrti committed:
wip

kashbrti committed:
wip/wasm-api

kashbrti committed:
add the vk generation changes to ts side

kashbrti committed:
wip

kashbrti committed:
cleanup, removing prints

kashbrti committed:
more cleanup

kashbrti committed:
more cleanup

kashbrti committed:
returning some needed prints

kashbrti committed:
updated prover tomls

kashbrti committed:
revert changes to end to end tests

kashbrti committed:
remove the filebase flow test

kashbrti committed:
Merge branch 'merge-train/barretenberg' into kb/bbapi_changes_for_hiding_circuit

kashbrti committed:
removing compute_civc_vk method

kashbrti committed:
Merge branch 'kb/bbapi_changes_for_hiding_circuit' of github.com:AztecProtocol/aztec-packages into kb/bbapi_changes_for_hiding_circuit

kashbrti committed:
Merge branch 'merge-train/barretenberg' of github.com:AztecProtocol/aztec-packages into merge-train/barretenberg

kashbrti committed:
Merge branch 'merge-train/barretenberg' into kb/bbapi_changes_for_hiding_circuit

kashbrti committed:
added hiding kernel to the number of circuits in benchmarks

ledwards2225 commented on noir-projects/client_ivc_tail_circuits.json:
```diff
@@ -1,4 +1 @@
-  "private_kernel_tail",
-  "private_kernel_tail_to_public"
-]
+["hiding"]
```
Hiding is already included in the more general "client_ivc_circuits"? Can we just get rid of this file altogether? @LeilaWang

LeilaWang replied:
This file is for generating the ivc vk:
`write_vk --scheme client_ivc --verifier_type ivc`

And the circuits defined in `client_ivc_circuits` have their vks generated as:
`write_vk --scheme client_ivc --verifier_type standalone`

Currently the hiding circuit has these 2 vks, and we copy the first to the `bb-prover` package to verify the civc proof. Looks like the second one is not used anywhere.

Do both commands produce the same vk for the hiding circuit? If so, then we can get rid of this file. But there will be some cleaning up to do in `bb-prover`. I can do it altogether in a follow up pr!

kashbrti replied:
The plan is that soon they will. But at the moment, one also includes an `ECCVM` and a `Translator` vk.

LeilaWang replied:
Aw then we need to keep this file for now. But I'll remove hiding from `client_ivc_circuits.json`!

ledwards2225 commented on barretenberg/cpp/src/barretenberg/api/api_client_ivc.cpp:
```diff
@@ -63,52 +63,28 @@ void write_standalone_vk(const std::string& output_format,
-} // anonymous namespace
+    return civc_vk_bytes;
+} // annonymus namespace
+
```
these namespace comments got a bit garbled. delete here and update subsequent line

kashbrti replied:
hmm, I tried putting the `// annonymus namespace` after we close the scope of the namespace but the formatter automatically changes it to `// namespace`

ledwards2225 commented on barretenberg/cpp/src/barretenberg/api/api_client_ivc.cpp:
```diff
@@ -118,19 +94,24 @@ void ClientIVCAPI::prove(const Flags& flags,
+        .circuit = { .name = step.function_name, .bytecode = step.bytecode, .verification_key = step.vk }
+    }.execute(request);
+    bbapi::ClientIvcHidingKernel{ .witness = step.witness }.execute(request);
 
```
The plan is to get rid of the need for this right?

kashbrti replied:
yep, if we get the hiding circuit completion to be done through the accumulate call this and all changes I've done to the API could go away

ledwards2225 commented on barretenberg/cpp/src/barretenberg/api/api_client_ivc.test.cpp:
```diff
@@ -124,49 +124,50 @@ ClientIVC::MegaVerificationKey get_ivc_vk(const std::filesystem::path& test_dir)
-    // Test verify command
-    EXPECT_TRUE(verify_proof());
-}
+// TEST_F(ClientIVCAPITests, ProveAndVerifyFileBasedFlow)
```
Use DISABLED_* instead. (I think this test was already disabled by Mara in a recent PR)

ledwards2225 commented on barretenberg/cpp/src/barretenberg/bbapi/bbapi_client_ivc.cpp:
```diff
@@ -117,37 +137,7 @@ static std::shared_ptr<ClientIVC::DeciderProvingKey> get_acir_program_decider_pr
-ClientIvcComputeStandaloneVk::Response ClientIvcComputeStandaloneVk::execute(BB_UNUSED const BBApiRequest& request) &&
+ClientIvcComputeStandaloneVk::Response ClientIvcComputeStandaloneVk::execute(const BBApiRequest& request) &&
 {
     info("ClientIvcComputeStandaloneVk - deriving VK for circuit '", circuit.name, "'");
```
It looks like this is already not being used anywhere - can we delete it now?

kashbrti replied:
`write_standalone_vk` uses it. which is used in the `write_vk` flow if we're not doing it for an IVC.

kashbrti replied:
I guess you meant the `ComputeCivcVK`. That is now being used in `write_civc_vk`. I made the changes so that WASM would do the same as native.

ledwards2225 commented on yarn-project/ivc-integration/src/native_client_ivc_integration.test.ts:
```diff
@@ -27,6 +27,7 @@ describe('Client IVC Integration', () => {
   // 3. Run the tail kernel to finish the client IVC chain.
+  // 4. Run the hiding kernel.
   it('Should generate a verifiable client IVC proof from a simple mock tx', async () => {
     const [bytecodes, witnessStack, _, vks] = await generate3FunctionTestingIVCStack();
```
Can you rename these test methods from generate3FunctionTestingIVCStack/generate6FunctionTestingIVCStack to something more general

ledwards2225 commented on yarn-project/ivc-integration/src/wasm_client_ivc_integration.test.ts:
```diff
@@ -103,68 +89,8 @@ describe('Client IVC Integration', () => {
   // 5. Run the reset kernel to process the read request emitted by the reader app
   // 6. Run the tail kernel to finish the client IVC chain
   it('Should generate a verifiable client IVC proof from a complex mock tx', async () => {
-    const tx = {
```
nice!

ledwards2225 commented on yarn-project/pxe/src/private_kernel/private_kernel_execution_prover.ts:
```diff
@@ -308,6 +298,38 @@ export class PrivateKernelExecutionProver {
 
+    // Hiding circuit is only executed if we are generating witnesses.
+    // For simulation, we can end with the tail, since the hiding circuit will simply return the same tail output.
+    if (generateWitnesses) {
```
I don't fully understand the context here but just want to make sure this isn't a brittle/implicit assumption that's likely to break in the future

LeilaWang replied:
We can skip the hiding circuit if the flag is not set, because this function is also used for simulating a transaction, to get the gas used and compute the fees, or to see what note hashes or logs etc have been created. So in that case, we don't need to run the hiding circuit. Because the users only want to see the values in the tail's public inputs.

kashbrti committed:
fixing merge mistake

ledwards2225 reviewed (approved):
Nice! Just a few cleanup requests. Looking forward to the follow up that cleans this up even more but happy with how simple the changes are so far!

kashbrti committed:
addressing review comments

kashbrti commented:
@LeilaWang I've addressed the comments on barretenberg side. If you'd like to give it a pass and address the last couple of comments we can merge this one in :)

kashbrti committed:
typo fix

LeilaWang committed:
Create util to flexibly generate testing ivc stack.

LeilaWang committed:
Update config.
